### PR TITLE
Fix Truffle invariants for Nothing with a warning

### PR DESCRIPTION
**Pull request by Akirathan**
This PR ensures that from the Truffle interop perspective, `Nothing` does not have meta parents.

### CODE REVIEW

File: engine/runtime/src/test/java/org/enso/interpreter/test/MetaObjectTest.java [hunk 1]
```diff
@@ -215,6 +217,25 @@ public void typesOfConstructors() throws Exception {
...
+  @Test
+  public void nothingWithWarningIsNotMeta() {
+    var src = """
+import Standard.Base.Warning.Warning
+import Standard.Base.Nothing.Nothing
+
+main = Warning.attach "foo" Nothing
+""";
+    var nothingWithWarn = TestBase.evalModule(ctx, src);
+    assertThat(nothingWithWarn.isMetaObject(), is(false));
```
Comment (Akirathan) [inline engine/runtime/src/test/java/org/enso/interpreter/test/MetaObjectTest.java:237]: @JaroslavTulach This test case fails with
```
[error] Test org.enso.interpreter.test.MetaObjectTest.nothingWithWarningIsNotMeta failed: org.graalvm.polyglot.PolyglotException: java.lang.AssertionError: Invariant contract violation for receiver Nothing(org.enso.interpreter.runtime.data.Type)., took 3.592 sec
[error]     at com.oracle.truffle.api.interop.InteropLibrary$Asserts.getMetaParents(InteropLibrary.java:5209)
[error]     at com.oracle.truffle.api.interop.InteropLibraryGen.genericDispatch(InteropLibraryGen.java:498)
[error]     at com.oracle.truffle.api.library.ReflectionLibraryDefault$Send.doSendCached(ReflectionLibrary.java:182)
[error]     at com.oracle.truffle.api.library.ReflectionLibraryDefaultGen$ReflectionLibraryExports$Cached.executeAndSpecialize(ReflectionLibraryDefaultGen.java:140)
[error]     at com.oracle.truffle.api.library.ReflectionLibraryDefaultGen$ReflectionLibraryExports$Cached.send(ReflectionLibraryDefaultGen.java:106)
[error]     at com.oracle.truffle.api.library.ReflectionLibraryGen$CachedDispatch.send(ReflectionLibraryGen.java:283)
[error]     at org.enso.interpreter.runtime.error.WithWarnings.send(WithWarnings.java:168)
[error]     at org.enso.interpreter.runtime.error.WithWarningsGen$ReflectionLibraryExports$Cached.send(WithWarningsGen.java:402)
[error]     at com.oracle.truffle.api.interop.InteropLibraryGen$Proxy.getMetaParents(InteropLibraryGen.java:3808)
[error]     at com.oracle.truffle.api.interop.InteropLibrary$Asserts.assertNoMetaObject(InteropLibrary.java:5094)
[error]     at com.oracle.truffle.api.interop.InteropLibrary$Asserts.isMetaObject(InteropLibrary.java:5070)
[error]     at com.oracle.truffle.polyglot.PolyglotValueDispatch$InteropValue$IsMetaObjectNode.doCached(PolyglotValueDispatch.java:4854)
[error]     at com.oracle.truffle.polyglot.PolyglotValueDispatchFactory$InteropValueFactory$IsMetaObjectNodeGen.executeAndSpecialize(PolyglotValueDispatchFactory.java:11021)
[error]     at com.oracle.truffle.polyglot.PolyglotValueDispatchFactory$InteropValueFactory$IsMetaObjectNodeGen.executeImpl(PolyglotValueDispatchFactory.java:10969)
[error]     at com.oracle.truffle.polyglot.HostToGuestRootNode.execute(HostToGuestRootNode.java:124)
[error]     at org.graalvm.polyglot.Value.isMetaObject(Value.java:219)
[error]     at org.enso.interpreter.test.MetaObjectTest.nothingWithWarningIsNotMeta(MetaObjectTest.java:237)
[error]     at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
[error]     at java.lang.reflect.Method.invoke(Method.java:580)
[error]     ...
```

(Run it with `sbt:runtime> testOnly *MetaObjectTest -- -z org.enso.interpreter.test.MetaObjectTest.nothingWithWarningIsNotMeta`).

The test case above succeeds. Shouldn't we fix this as well?

Comment (Akirathan) [reply engine/runtime/src/test/java/org/enso/interpreter/test/MetaObjectTest.java:237]: Note that this test case is a minimal reproducer of https://github.com/enso-org/enso/pull/8595/files/818c1ef2910c9f1e5608b2aee4da162a604d2fe1..8b685a136549fb11809ab4d8f23f171192688ba3. I have debugged the internal Truffle code and saw that for some reason, it calls `isMetaObject` on a Nothing with a warning attached.

Comment (Akirathan) [reply engine/runtime/src/test/java/org/enso/interpreter/test/MetaObjectTest.java:237]: cc @hubertp

Comment (Akirathan) [reply engine/runtime/src/test/java/org/enso/interpreter/test/MetaObjectTest.java:237]: Answer to myself: Yes, that should not throw `AssertionError`. The `isMetaObject` on that object should return just false.

Comment (JaroslavTulach) [reply engine/runtime/src/test/java/org/enso/interpreter/test/MetaObjectTest.java:237]: We certainly shouldn't be getting:

> java.lang.AssertionError: Invariant contract violation for receiver Nothing

but rather `false` or no answer, etc.

**Comment by Akirathan** [comment]
Timeout in `EngineManagementApiSpec`: https://github.com/enso-org/enso/actions/runs/7743955600/job/21116569741?pr=8764#step:8:1640

<blockquote><div> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">Fix Truffle invariants for Nothing with a warning · enso-org/enso@ad5ec40</a></strong></div><div>Hybrid visual and textual functional programming. Contribute to enso-org/enso development by creating an account on GitHub.</div></blockquote>

File: engine/runtime/src/test/java/org/enso/interpreter/test/MetaObjectTest.java [hunk 2]
```diff
@@ -1,5 +1,7 @@
 package org.enso.interpreter.test;
 
+import static org.hamcrest.CoreMatchers.*;
```
Comment (JaroslavTulach) [inline engine/runtime/src/test/java/org/enso/interpreter/test/MetaObjectTest.java]: Please avoid `*` wildcard imports.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/Type.java [hunk 3]
```diff
@@ -196,15 +196,18 @@ Type getMetaObject(@CachedLibrary("this") InteropLibrary lib) throws Unsupported
   @ExportMessage
   Object getMetaParents(@CachedLibrary("this") InteropLibrary lib)
       throws UnsupportedMessageException {
-    if (isNothing(lib) || !hasMetaParents()) {
+    if (isNothing(lib) || !hasMetaParents(lib)) {
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/data/Type.java]: This is calling `isNothing` and also `hasMetaParents`, but `hasMetaParents` is calling `isNothing`. I cannot make my head around this.

**Review by JaroslavTulach — COMMENTED** [review]
Test is fine. The implementation feels strange... or I am not getting it.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/Type.java (hunk 3 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/data/Type.java]: Shouldn't this check be just:
```
if (!hasMetaParents(lib)) {
  throw UnsupportMessageException.create();
}
```
?

File: engine/runtime/src/test/java/org/enso/interpreter/test/MetaObjectTest.java (hunk 2 above)
Comment (Akirathan) [reply engine/runtime/src/test/java/org/enso/interpreter/test/MetaObjectTest.java]: Sorry. Fixed in 44c834291dcb71c5798dc02c060bd52604aadf74. Apparently, our `javafmt` command does not replace star imports, I will have to be more careful about this in the future.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/Type.java (hunk 3 above)
Comment (Akirathan) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/data/Type.java]: Yeah, good observation.  The `isNothing` call is now redundant. Removed that in e01e85ab35452b18b96d68ccbf64df207cdc8a4d

**Review by JaroslavTulach — APPROVED** [review]

**Pull request merged**